q2_diversity_lib/tests/test_util.py

```diff
@@ -158,3 +174,40 @@ def test_auto_passed_to_cpu_request(self, mock_process):
         self.assertEqual(self.function_w_n_jobs_param(n_jobs='auto'), 3)
         self.assertEqual(self.function_w_threads_param('auto'), 3)
         self.assertEqual(self.function_w_threads_param(threads='auto'), 3)
+
+    @mock.patch("q2_diversity_lib._util.psutil.Process")
```

Comment:
> HMMM: is stderror breaking GHActions?

No - unifrac is breaking because of the patching in this test. The patch is explicitly telling q2-div-lib to detect three CPUs, (which will make `auto` translate to `3`. This is what happens on my machine (hardware will matter here, because unifrac does its own hardware interrogation, I assume):

```
Unable to bind this thread to core 36450304. Are sufficient processors available?Unable to bind thread 2 to core: -1
```

(btw - I had to tweak my invocation of py.test to see that, since unifrac straight up bombs:

```bash
py.test --capture=tee-sys -k test_cpu_request_through_framework
```

I think its okay to not worry about patching in these tests, since its really just about exercising that all the independently verified parts work together. In that case, my vote is to drop the patching, and test n_jobs/threads at `auto`, and `1` (rather than `2`).

Reply:
> hardware will matter here

To elaborate on that, my machine simply doesn't have that many CPUS:

```python
diversity_lib.methods.unweighted_unifrac(table=table, phylogeny=tree, threads=3)                                                                                                                                                                                                                        
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-bed7731ab85d> in <module>
----> 1 diversity_lib.methods.unweighted_unifrac(table=table, phylogeny=tree, threads=3)

<decorator-gen-148> in unweighted_unifrac(table, phylogeny, threads, bypass_tips)

~/projects/qiime2/qiime2/qiime2/sdk/action.py in bound_callable(*args, **kwargs)
    243                 # Execute
    244                 outputs = self._callable_executor_(scope, callable_args,
--> 245                                                    output_types, provenance)
    246 
    247                 if len(outputs) != len(self.signature.outputs):

~/projects/qiime2/qiime2/qiime2/sdk/action.py in _callable_executor_(self, scope, view_args, output_types, provenance)
    388 
    389     def _callable_executor_(self, scope, view_args, output_types, provenance):
--> 390         output_views = self._callable(**view_args)
    391         output_views = tuplize(output_views)
    392 

<decorator-gen-133> in unweighted_unifrac(table, phylogeny, threads, bypass_tips)

~/projects/qiime2/q2-diversity-lib/q2_diversity_lib/_util.py in _disallow_empty_tables(wrapped_function, *args, **kwargs)
     51         raise ValueError("The provided table is empty")
     52 
---> 53     return wrapped_function(*args, **kwargs)
     54 
     55 

<decorator-gen-132> in unweighted_unifrac(table, phylogeny, threads, bypass_tips)

~/projects/qiime2/q2-diversity-lib/q2_diversity_lib/_util.py in _validate_requested_cpus(wrapped_function, *args, **kwargs)
     90 
     91     if cpus_requested > cpus_available:
---> 92         raise ValueError(f"The value passed to '{param_name}' cannot exceed "
     93                          f"the number of processors ({cpus_available}) "
     94                          "available to the system.")

ValueError: The value passed to 'threads' cannot exceed the number of processors (2) available to the system.
```

Reply:
Thanks again. `py.test` has a whole world of flags I'm not familiar with yet. `capture` is what I've been looking for. :mag:

Reply:
pytest:

![image](https://user-images.githubusercontent.com/274668/84547621-ce1a2000-acb8-11ea-86a0-97c9b1aadc49.png)

Reply:
https://youtu.be/e-JHfXVlkik